src/Compilers/Core/Portable/Operations/Operation.cs

```diff
@@ -46,41 +76,203 @@ public Operation(OperationKind kind, SyntaxNode syntax, ITypeSymbol type, Option
+            Interlocked.Exchange(ref (operation as Operation)._parentDoNotAccessDirectly, null);
+            return operation;
+        }
 
-            public Optional<object> ConstantValue { get; }
+        private class NoneOperation : Operation
+        {
+            private readonly Func<ImmutableArray<IOperation>> _getChildren;
 
-            public void Accept(OperationVisitor visitor)
+            public NoneOperation(SemanticModel semanticMode, SyntaxNode node, Optional<object> constantValue, Func<ImmutableArray<IOperation>> getChildren) :
+                base(OperationKind.None, semanticMode, node, type: null, constantValue: constantValue)
+            {
+                _getChildren = getChildren;
+            }
+
+            public override void Accept(OperationVisitor visitor)
             {
                 visitor.VisitNoneOperation(this);
             }
 
-            public TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument)
+            public override TResult Accept<TArgument, TResult>(OperationVisitor<TArgument, TResult> visitor, TArgument argument)
             {
                 return visitor.VisitNoneOperation(this, argument);
             }
 
-            public IEnumerable<IOperation> Children => _getChildren().NullToEmpty();
+            public override IEnumerable<IOperation> Children
+            {
+                get
+                {
+                    foreach (var child in _getChildren().NullToEmpty().WhereNotNull())
+                    {
+                        yield return Operation.SetParentOperation(child, this);
+                    }
+                }
+            }
+        }
+
+        private static readonly ObjectPool<Queue<IOperation>> s_queuePool =
+            new ObjectPool<Queue<IOperation>>(() => new Queue<IOperation>(), 10);
+
+        private IOperation WalkDownOperationToFindParent(HashSet<IOperation> operationAlreadyProcessed, IOperation root)
+        {
+            void EnqueueChildOperations(Queue<IOperation> queue, IOperation parent)
+            {
+                // for now, children can return null. once we fix the issue, children should never return null
+                // https://github.com/dotnet/roslyn/issues/21196
+                foreach (var o in parent.Children.WhereNotNull())
+                {
+                    queue.Enqueue(o);
+                }
+            }
+
+            var operationQueue = s_queuePool.Allocate();
+
+            try
+            {
+                EnqueueChildOperations(operationQueue, root);
+
+                // walk down the tree to find parent operation
+                // every operation returned by the queue should already have Parent operation set
+                while (operationQueue.Count > 0)
+                {
+                    var operation = operationQueue.Dequeue();
+
+                    if (!operationAlreadyProcessed.Add(operation))
```

Comment:
since we move up the spine through syntax node to find child, current operation tree will be always sub tree of new tree we found from the syntax node. no need to search again current tree since the reason we moved up is because we didn't find one in current tree.